csharp/src/Microsoft.ML.OnnxRuntime/NamedOnnxValue.shared.cs

```diff
@@ -18,16 +18,19 @@ namespace Microsoft.ML.OnnxRuntime
     /// </summary>
     internal class MapHelper
     {
-        internal MapHelper(object keys, object values)
+        internal MapHelper(TensorBase keys, TensorBase values)
         {
             Keys = keys;
             Values = values;
         }
-        internal Object Keys { get; }   // DenseTensor<K>
-        internal Object Values { get; } // DenseTensor<V>
+        internal TensorBase Keys { get; }   // DenseTensor<K>
+        internal TensorBase Values { get; } // DenseTensor<V>
     }
 
     /// <summary>
+    /// This is a legacy class that is kept for backward compatibility.
+    /// Use OrtValue based API.
```

Comment:
Just like with C++ API the names are kept in a parallel structure, but I am open for suggestions. The fact that NamedOnnxValue has a name is probably the only positive here.

Reply:
What parallel structure contains the name if NamedOnnxValue is a legacy class that may go away?

FWIW this approach mirrors the ONNX protobuf setup - you have values (Tensor/Sequence/Map/etc) and ValueInfo which  is the structure which adds a name to a value. That's also why a sequence only has a name for the overall sequence and not individual values in it.

Reply:
We need to chat on this one. I am failing to understand why the approach we take in C++ would not work here.

`ValueInfo` is a metadata that has `NodeMetadata` structure equivalent, and `NamedOnnxValue` and `OrtValue` are runtime structures. When `NamedOnnxValue` is used to host nested types it does not have to have a name.
Likewise, `OrtValue` generally does not have a name, but input and output names are supplied as a separate arguments in the `Run*()` calls.

Reply:
In Java we don't name the OrtValues at all, the names are supplied separately from maps, because the name is a properly of an input or output point rather than the tensor itself. I don't understand how the C# code currently works if you take an output tensor with one name and pass it into a different model with a different name.

Reply:
`NamedOnnxValue` and other types alongside is a poor legacy choice. I wasted lots of time supporting it, while the onl[y thing we need is `OrtValue`.

Reply:
Got it. Something similar to the Java setup would be nice - i.e. use a map for input/output to create the association between an OrtValue and a name as that's the only place it's needed. An OrtValue that contains a sequence should contain a collection of OrtValue instances for the sequence items. All the *OnnxValue stuff gets deprecated. 

I don't get the need to have separate arguments for the names vs the values for input and output in InferenceSession::Run though. Whilst the C API has to do that, at the C# level a Dictionary<string, OrtValue> would make more sense wouldn't it?

Reply:
I am not sure I follow.